pkgs/build-support/singularity-tools/default.nix

```diff
@@ -109,14 +81,280 @@ rec {
+  #         }
+  #         ''
+  #           rm -rf $out
+  #           mkdir disk
+  #           mkfs -t ext3 -b 4096 /dev/${vmTools.hd}
+  #           mount /dev/${vmTools.hd} disk
+  #           mkdir -p disk/img
+  #           cd disk/img
+  #           mkdir proc sys dev
 
+  #           # Run root script
+  #           ${lib.optionalString (runAsRoot != null) ''
+  #             mkdir -p ./${storeDir}
+  #             mount --rbind ${storeDir} ./${storeDir}
+  #             unshare -imnpuf --mount-proc chroot ./ ${runAsRootFile}
+  #             umount -R ./${storeDir}
+  #           ''}
+
+  #           # Build /bin and copy across closure
+  #           mkdir -p bin ./${storeDir}
+  #           for f in $(cat $layerClosure) ; do
+  #             cp -ar $f ./$f
+  #           done
+
+  #           for c in ${toString contents} ; do
+  #             for f in $c/bin/* ; do
+  #               if [ ! -e bin/$(basename $f) ] ; then
+  #                 ln -s $f bin/
+  #               fi
+  #             done
+  #           done
+
+  #           # Create runScript and link shell
+  #           if [ ! -e bin/sh ]; then
+  #             ln -s ${runtimeShell} bin/sh
+  #           fi
+  #           mkdir -p .${projectName}.d
+  #           ln -s ${runScriptFile} .${projectName}.d/runscript
+
+  #           # Fill out .${projectName}.d
+  #           mkdir -p .${projectName}.d/env
+  #           touch .${projectName}.d/env/94-appsbase.sh
+
+  #           cd ..
+  #           mkdir -p /var/lib/${projectName}/mnt/{container,final,overlay,session,source}
+  #           echo "root:x:0:0:System administrator:/root:/bin/sh" > /etc/passwd
+  #           echo > /etc/resolv.conf
+  #           TMPDIR=$(pwd -P) ${projectName} build $out ./img
+  #         '');
+
+  #   in
+  #   result;
+
+  inherit (import ./definition-lib.nix { inherit lib; })
+    knownPrimarySectionNamesDefault
+    knownAppSectionNamesDefault
+    toSingularityDef
+    ;
+
+  contentsToDef =
+    { contents ? [ ]
+    , definitionOverrider ? null
+    }:
+    let
+      layerClosure = writeMultipleReferencesToFile (contents ++ [ bash coreutils ]);
+    in
+    (
+      if lib.isFunction definitionOverrider then
+        definitionOverrider
+      else if builtins.isAttrs definitionOverrider then
+        (d: lib.recursiveUpdate d definitionOverrider)
+      else
+        lib.id
+    ) {
+      header.Bootstrap = "scratch";
+      setup = ''
+        mkdir -p ''${SINGULARITY_ROOTFS}/${storeDir}
+        for f in $(cat ${layerClosure}) ; do
+          cp -r "$f" "''${SINGULARITY_ROOTFS}/${storeDir}"
+        done
+        mkdir -p "''${SINGULARITY_ROOTFS}/bin"
+        "${coreutils}/bin/ln" -s "${runtimeShell}" "''${SINGULARITY_ROOTFS}/bin/sh"
+        mkdir -p "''${SINGULARITY_ROOTFS}/usr/bin"
+        "${coreutils}/bin/ln" -s "${coreutils}/bin/env" "''${SINGULARITY_ROOTFS}/usr/bin/env"
+      '';
+      environment = {
+        PATH = "${lib.makeBinPath contents}:\${PATH:-}";
```

Comment:
I suppose this the alternative to the previous approach of generating a FHS-like tree (`/bin` and such). We can address this in a separate PR, but I wanted to explore (ab)using ext3 and populate both `/nix/store` and `/bin` using hardlinks. This way I could hide the container's `/nix` with `--bind /tmp:/nix` and still having working `sh` and `nix` (assuming I've built them statically linked)

Reply:
> I suppose this is the alternative to the previous approach of generating a FHS-like tree (`/bin` and such).

Yes. The `PATH`-based approach seems cleaner for me, which is also how `nix shell` works. As both of them has its use cases (Nixy idiom vs. FHS PATH compatibility), we could allow users to decide which to use.



> We can address this in a separate PR, but I wanted to explore (ab)using ext3 and populate both `/nix/store` and `/bin` using hardlinks. This way I could hide the container's `/nix` with `--bind /tmp:/nix` and still having working `sh` and `nix` (assuming I've built them statically linked)

HPC might be based upon a distributed filesystems; great chance that they don't support hard links. Former versions of Singularity simply extracts the SquashFS image on the host. If the machine have some sort of TMPDIR where local filesystems like ext4 or tmpfs are mounted, you could have it extract there and (hopefully) enjoy the hard links. If you don't have access hard-link supported spaces, there might be problems extracting an image like that.

A good news is that both Apptainer and SingularityCE now tries to mount the image with the `squashfuse` command from the environment before falling back to extracting, and that SquashFS does support hard links on its own. I added `squashfuse`  to the defaultPath and wrapper PATH of both `apptainer` and `singularity` along with the version bump #224683. Let's make it land on the stable release of Nixpkgs.

Reply:
> great chance that they don't support hard links

I guess I don't really care if they're extracted as copies on the host, as long as they are represented as hardlinks inside the image that we create.

>  I added squashfuse to the defaultPath and wrapper PATH of both apptainer and singularity along with the version bump

Awesome! I should have a look tomorrow!

> Let's make it land on the stable release of Nixpkgs

Sounds great! But just in case it takes longer, we can and should backport later